### PR TITLE
add on-cluster healtcheck test and prevent enforced false on not healhy probes

### PR DESCRIPTION
**Pull request by maksymvavilov**

### CODE REVIEW

File: pkg/library/kuadrant/apimachinery_status_conditions.go [hunk 1]
```diff
@@ -78,3 +81,27 @@ func EnforcedCondition(policy Policy, err PolicyError, fully bool) *metav1.Condi
 
 	return cond
 }
+
+func HealthyCondition(policy Policy, err PolicyError) *metav1.Condition {
```
Comment (maleck13) [inline pkg/library/kuadrant/apimachinery_status_conditions.go:85]: not 100% sure if this should be here or not @mikenairn ?

File: tests/common/dnspolicy/dnspolicy_controller_test.go [hunk 2]
```diff
@@ -1244,6 +1244,102 @@ var _ = Describe("DNSPolicy controller", func() {
...
+							"Status":  Equal(metav1.ConditionTrue),
+							"Reason":  Equal(string(kuadrant.PolicyReasonEnforced)),
+							"Message": Equal("DNSPolicy has been successfully enforced"),
+						}),
+						MatchFields(IgnoreExtras, Fields{
+							"Type":   Equal(string(kuadrant.PolicyConditionSubResourcesHealthy)),
+							"Status": Equal(metav1.ConditionFalse),
+							"Reason": Equal(string(kuadrant.PolicyReasonUnknown)),
+							"Message": And(
+								ContainSubstring("DNSPolicy has encountered some issues: not all sub-resources of policy are healthy. Not healthy DNSRecords are:"),
```
Comment (maleck13) [inline tests/common/dnspolicy/dnspolicy_controller_test.go:1333]: should this message be 
` not all sub-resources of policy are passing the policy defined health check?` ? might be a bit more obvious then?

**Review by maleck13 — CHANGES_REQUESTED** [review]
We need these changes against v1 now. Also not sure if we want the condition stuff under library will wait for @mikenairn  on that one.  Otherwise it looks good

File: pkg/library/kuadrant/apimachinery_status_conditions.go (hunk 1 above)
Comment (mikenairn) [reply pkg/library/kuadrant/apimachinery_status_conditions.go:85]: Are you expecting this to ever be used by anything other than DNSPolicy? Seems unlikely, I'd probably move it all into the dns_workflow

File: tests/common/dnspolicy/dnspolicy_controller_test.go (hunk 2 above)
Comment (maksymvavilov) [reply tests/common/dnspolicy/dnspolicy_controller_test.go:1333]: changed the message

File: pkg/library/kuadrant/apimachinery_status_conditions.go (hunk 1 above)
Comment (maksymvavilov) [reply pkg/library/kuadrant/apimachinery_status_conditions.go:85]: It is strange to have a condition defined in the workflow and not alongside other conditions :thinking:  don't have a strong opinion on this, however - moved

File: api/v1alpha1/dnspolicy_types.go [hunk 3]
```diff
@@ -289,9 +289,9 @@ func (p *DNSPolicy) WithTargetGatewayListener(gwName string, lName string) *DNSP
 func (p *DNSPolicy) WithHealthCheckFor(endpoint string, port int, protocol string, failureThreshold int) *DNSPolicy {
 	return p.WithHealthCheck(dnsv1alpha1.HealthCheckSpec{
 		Path:             endpoint,
-		Port:             &port,
+		Port:             port,
```
[the remark this answers is not in the document]
Comment (maksymvavilov) [reply api/v1alpha1/dnspolicy_types.go:292]: rebased. It is v1 now

[the remark this answers is not in the document]
Comment (maleck13) [reply api/v1alpha1/dnspolicy_types.go:292]: no as in it needs go into the types in the v1 folde

[the remark this answers is not in the document]
Comment (maksymvavilov) [reply api/v1alpha1/dnspolicy_types.go]: I have a bad habit of commenting before the push. It is, I promise :laughing:

File: tests/common/dnspolicy/dnspolicy_controller_test.go [hunk 4]
```diff
@@ -1244,6 +1245,143 @@ var _ = Describe("DNSPolicy controller", func() {
...
+								"Type":    Equal(string(kuadrantdnsv1alpha1.ConditionTypeReady)),
+								"Status":  Equal(metav1.ConditionTrue),
+								"Reason":  Equal(string(kuadrantdnsv1alpha1.ConditionReasonProviderSuccess)),
+								"Message": Equal("Provider ensured the dns record"),
+							}),
+						),
+						Not(ContainElement(
+							MatchFields(IgnoreExtras, Fields{
+								"Type": Equal(string(kuadrantdnsv1alpha1.ConditionTypeHealthy)),
+							}),
```
Comment (maksymvavilov) [inline tests/common/dnspolicy/dnspolicy_controller_test.go:1345]: This one is redundant as well. If we are ready then we either don't have healthy cond or it is true. For this suite we never have `healthy - true` but I still would like to have a check for this

File: tests/common/dnspolicy/dnspolicy_controller_test.go [hunk 5]
```diff
@@ -1244,6 +1245,143 @@ var _ = Describe("DNSPolicy controller", func() {
...
+							"Message": Equal("DNSPolicy has been accepted"),
+						}),
+						MatchFields(IgnoreExtras, Fields{
+							"Type":   Equal(string(kuadrant.PolicyConditionEnforced)),
+							"Status": Equal(metav1.ConditionTrue),
+							"Reason": Equal(string(kuadrant.PolicyReasonEnforced)),
+							"Message": And(
+								ContainSubstring("DNSPolicy has been partially enforced. Not ready DNSRecords are:"),
+								ContainSubstring(recordName),
+								Not(ContainSubstring(wildcardRecordName))),
```
Comment (maksymvavilov) [inline tests/common/dnspolicy/dnspolicy_controller_test.go:1368]: If we are partially enforced with only two records it is sufficient to only check for the presence of one of the records. Leaving it here to be extra cautious.

**Pull request merged**